lib/archethic/mining/validation_context.ex

```diff
@@ -398,45 +393,107 @@ defmodule Archethic.Mining.ValidationContext do
     ...>     %Node{last_public_key: "key3"},
     ...>     %Node{last_public_key: "key4"}
+    ...>   ],
+    ...>   chain_storage_nodes: [
+    ...>     %Node{last_public_key: "key5"},
+    ...>     %Node{last_public_key: "key6"},
+    ...>     %Node{last_public_key: "key7"}
     ...>   ]
     ...> }
-    ...> |> ValidationContext.cross_validation_node?("key1")
+    ...> 
+    ...> ValidationContext.cross_validation_node?(context, "key3")
+    ...> true
+    ...> ValidationContext.cross_validation_node?(context, "key7")
+    ...> true
+    ...> ValidationContext.cross_validation_node?(context, "key1")
     false
+
   """
   @spec cross_validation_node?(t(), Crypto.key()) :: boolean()
   def cross_validation_node?(
-        %__MODULE__{cross_validation_nodes: cross_validation_nodes},
+        %__MODULE__{
+          cross_validation_nodes: cross_validation_nodes,
+          chain_storage_nodes: chain_storage_nodes
+        },
         node_public_key
       )
       when is_binary(node_public_key) do
-    Enum.any?(cross_validation_nodes, &(&1.last_public_key == node_public_key))
+    Enum.any?(
+      cross_validation_nodes ++ chain_storage_nodes,
+      &(&1.last_public_key == node_public_key)
+    )
   end
 
+  @doc """
+  Determines if enough cross validation stamps have been received to create the aggregated signature
+  Returns
+    - :reached if enough stamps are valid
+    - :not_reached if not enough stamps received yet
+    - :error if it's not possible to reach the threshold
+  """
+  @spec get_cross_validation_state(t()) :: :reached | :not_reached | :error
+  def get_cross_validation_state(%__MODULE__{
+        cross_validation_stamps: stamps,
+        sorted_nodes: sorted_nodes
+      }),
+      do: ProofOfValidation.get_state(sorted_nodes, stamps)
+
+  @doc """
+  Aggregate signature of the cross validation stamps
+  Create a bitmask of the node that signed the cross stamps
+  """
+  @spec create_proof_of_validation(context :: t()) :: t()
+  def create_proof_of_validation(
+        context = %__MODULE__{cross_validation_stamps: stamps, sorted_nodes: sorted_nodes}
+      ) do
+    %__MODULE__{context | proof_of_validation: ProofOfValidation.create(sorted_nodes, stamps)}
+  end
+
+  @doc """
+  Ensure a proof of validation is valid according to current context
+  """
+  @spec valid_proof_of_validation?(context :: t(), proof :: ProofOfValidation.t()) :: boolean()
+  def valid_proof_of_validation?(
+        context = %__MODULE__{
+          chain_storage_nodes: storage_nodes,
+          sorted_nodes: sorted_nodes,
+          validation_stamp: stamp
+        },
+        proof
+      ) do
+    validation_nodes = ProofOfValidation.get_nodes(sorted_nodes, proof)
+
+    expected_validation_nodes =
+      context |> get_confirmed_validation_nodes() |> Enum.concat(storage_nodes) |> Enum.uniq()
+
+    Enum.all?(validation_nodes, &Enum.member?(expected_validation_nodes, &1)) and
+      ProofOfValidation.valid?(sorted_nodes, proof, stamp)
```

Comment:
Maybe we should pass the elected nodes to the valid? function so it can check the validation nodes were part of the election.